In [1]:
import json
import requests
import pandas as pd
import numpy as np

from pandas import DataFrame

In [101]:
api_key = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJhY2NvdW50X2lkIjoiNTAzNDY4MjEzIiwiYXV0aF9pZCI6IjIiLCJ0b2tlbl90eXBlIjoiQWNjZXNzVG9rZW4iLCJzZXJ2aWNlX2lkIjoiNDMwMDExNDgxIiwiWC1BcHAtUmF0ZS1MaW1pdCI6IjIwMDAwOjEwIiwibmJmIjoxNTgxMDg0OTg5LCJleHAiOjE2NDQxNTY5ODksImlhdCI6MTU4MTA4NDk4OX0.tg1PMEcRjukDKo3SpCrbhXKQk82nIsKVXRv6LosKPro'

## 매치 종류

In [3]:
match_url = requests.get('https://static.api.nexon.co.kr/fifaonline4/latest/matchtype.json')
match_parsed_data = match_url.json()
match_type = pd.DataFrame(match_parsed_data)

In [4]:
match_type.head()

desc  matchtype
0  공식경기         50
1  감독모드         52

## 선수 고유 식별자

In [5]:
spid_url = requests.get('https://static.api.nexon.co.kr/fifaonline4/latest/spid.json')
spid_parsed_data = spid_url.json()
spid = pd.DataFrame(spid_parsed_data)

In [7]:
spid.head()

id       name
0  101000001    데이비드 시먼
1  101000195     로비 파울러
2  101000393     테디 셰링엄
3  101000805      옌스 레만
4  101001040  호베르투 카를루스

In [8]:
len(spid)

34095

게임에서 유저가 사용할 수 있는 선수의 수는 총 34095명이다. 선수는 ```id```로 관리된다.

## 선수 포지션

In [9]:
spposition_url = requests.get('https://static.api.nexon.co.kr/fifaonline4/latest/spposition.json')
spposition_parsed_data = spposition_url.json()
spposition = pd.DataFrame(spposition_parsed_data)

In [99]:
spposition

desc  spposition
0    GK           0
1    SW           1
2   RWB           2
3    RB           3
4   RCB           4
5    CB           5
6   LCB           6
7    LB           7
8   LWB           8
9   RDM           9
10  CDM          10
11  LDM          11
12   RM          12
13  RCM          13
14   CM          14
15  LCM          15
16   LM          16
17  RAM          17
18  CAM          18
19  LAM          19
20   RF          20
21   CF          21
22   LF          22
23   RW          23
24   RS          24
25   ST          25
26   LS          26
27   LW          27
28  SUB          28

한 선수는 GK부터 SUB까지 총 29개의 포지션을 가질 수 있으며, 포지션은 ```spposition```으로 관리된다.

## 시도한 방법 1

기준을 <b>선수</b>로 하여, 선수 1명당 가질 수 있는 포지션의 경우를 모두 계산해 데이터프레임으로 저장한다. 

In [104]:
a = ['a', 'b', 'c', 'd', 'e']
b = [1,2,3]

df = pd.DataFrame(columns = ['선수id', '포지션'])
for i in a :
    for j in b :
        df = df.append(pd.DataFrame([[i, j]], columns = ['선수id', '포지션']))

df

선수id 포지션
0    a   1
0    a   2
0    a   3
0    b   1
0    b   2
0    b   3
0    c   1
0    c   2
0    c   3
0    d   1
0    d   2
0    d   3
0    e   1
0    e   2
0    e   3

In [ ]:
df = pd.DataFrame(columns = ['id', 'po'])

for i in spid['id'] :
    for j in spposition['spposition'] :
        df = df.append(pd.DataFrame([[i, j]], columns = ['id', 'po']))

df

### 문제 1
모든 선수들이 가질 수 있는 포지션의 경우의 수를 모두 데이터 프레임에 추가하려고 했는데, 실행 시간이 너무 오래 걸린다.

### 해결 1

선수들의 수를 10명씩 나누어서 데이터 프레임을 만든다.

In [93]:
def divide_list(l, n):
    for i in range(0, len(l), n) :
        yield l[i:i+n]

n = 10

spid_list = list(divide_list(spid['id'], n))

In [94]:
len(spid_list)

3410

In [91]:
def df_maker(l) :
    df = pd.DataFrame(columns = ['id', 'po'])
    for i in spid_list[l] :
        for j in spposition['spposition'] :
            df = df.append(pd.DataFrame([[i, j]], columns = ['id', 'po']))

### 문제 1-1
결국 실행해야 하는 반복문의 수가 너무 많다!

## API 조회하기

TOP 10,000 랭커 유저가 사용한 선수의 20경기 평균 스탯을 조회합니다.
선수의 고유 식별자와 포지션의 목록으로 랭커 유저가 사용한 선수의 평균 스탯 기록을 조회할 수 있습니다.
한번에 너무 많은 선수목록을 입력할 경우 413 Request Entity Too Large 에러가 반환될 수 있습니다.
한번에 호출하는 선수의 수는 50명이 적당합니다.

조회하고자 하는 선수 목록을 다음과 같이 구성하여 {players} 파라미터로 전송하면 API를 조회할 수 있습니다 :

{players} : "id", "po"필드를 가지고 있는 Json Object Array
- id : 선수 고유 식별자 (spid, /metadata/spid API 참고)
- po : 선수 포지션 (spposition, /metadata/spposition API 참고)

ex) [{"id":101001183,"po":7}, {"id":214003647,"po":25},…]

*데이터는 매일 3, 9, 15, 21시에 갱신을 시작하며 최대 한시간까지 소요될 수 있습니다.
갱신된 데이터는 갱신시작 시점을 기준으로 3시간 전 데이터까지 반영됩니다.


In [103]:
headers = {'Authorization' : api_key}
r = requests.get("https://api.nexon.co.kr/fifaonline4/v1.0/rankers/status?matchtype=50&players=[{'id': 101001183, 'po': 1}]", headers = headers)

r.content

b'[{"spId":101001183,"spPosition":1,"status":{"shoot":0.0,"effectiveShoot":0.0,"assist":0.0,"goal":0.0,"dribble":0.0,"passTry":3.66667,"passSuccess":3.66667,"block":0.0,"tackle":1.33333,"matchCount":3},"createDate":"2020-02-13T02:30:00"}]'

http://hleecaster.com/python-how-to-split-a-list-into-chunks/    
https://programmers.co.kr/learn/courses/4008/lessons/12738     
https://specialmylife.tistory.com/entry/pandas-DataFrame-%EC%82%AC%EC%9A%A9%EB%B2%95-%EC%A0%95%EB%A6%AC%EC%9E%91%EC%97%85%EC%A4%91